reactor-core/src/main/java/reactor/core/publisher/MonoZip.java

```diff
@@ -333,12 +339,18 @@ public void onSubscribe(Subscription s) {
 		public void onNext(Object t) {
 			if (value == null) {
 				value = t;
-				parent.signal();
+				if (!parent.signal()) {
```

Comment:
false means the signal was not delivered which happens when the process was canceled or force terminated with the error or empty source. In that scenario value is discarded by a **caller** since the `signal` method has no information about the **caller** or signaled **value**

Reply:
The way it is implemented right now seems different. Also, please have a look at failing tests from CI.
The condition for returning false is:

            if (DONE.incrementAndGet(this) != n) {
 				return false;
 			}

And that is the same condition which signifies the zipping is done. I'm a bit confused.

Reply:
right

Reply:
fixed